Open CV Docker Image Development

theodorelu27 opened the pull request:
Initial Pull Request for #25 

Included are:
- Dockerfile to build an Ubuntu client with Java8 and OpenCV 4.5
- Sample codes for marker creation and detection
note: sample codes were added in the ``samples` directory to exclude it during build

Missing:
- Dockerfile not yet working
- Connection between nodes and server

theodorelu27 commented:
@alyhasansakr I've found out the problem for the Docker image build, it's due to my Dockerfile being saved with CRLF (Windows End of File) instead of LF (Linux End of File)

so the build is working as intended now, I'd focus on the client-server connection now

theodorelu27 commented:
@alyhasansakr I've pushed my most recent version

What has been done:
- Dockerfile to build an Ubuntu 16, OpenJDK8, OpenCV4.5 (Ubuntu and OpenJDK version may be changed)
- CreateMarker and DetectMarker
- Client that can be configured when running the class to use server or run locally
- Image File transfer between Client and Server over TCP using BufferedImage and ImageIO
- Server that handles and send back the detected marker image

What's still missing:
- Thread execution
- Master communication thread
- External configuration
- Configuration for cmake to reduce build time
- Unit testing

I haven't finished refactoring the sample codes and as of currently, the cmake still builds everything and takes around 1 hour to finish

I've also added the Test Files in a zip file in the samples directory which includes the image used in the test, as well as a pre-built opencv450 for Linux to cut down build time

alyhasansakr commented on samples/OpenCVClient/CreateMarker.java:
```diff
@@ -0,0 +1,28 @@
+import java.util.Vector;
+
+public class CreateMarker {
+   public static void main(String[] args) {
```
I am with you that we need another a separate application to make the markers, but the location is not ideal, it should not be a part of the client.

My idea is that we use this "MarkerGenerator" to make images that we later use with the OpenCVClient. So images should be copied to the docker image that runs the OpenCVClient, and MarkerGenerator should not a part of that docker image.

alyhasansakr commented on samples/OpenCVServer/CreateMarker.java:
```diff
@@ -0,0 +1,28 @@
+import java.util.Vector;
+
+public class CreateMarker {
+   public static void main(String[] args) {
```
See my other comment in OpenCVClient/CreateMarker.java

alyhasansakr commented on samples/OpenCVServer/OpenCVServer.java:
```diff
@@ -0,0 +1,52 @@
+            ImageIO.write(currentImage, "png", new File(fileName));
+            DetectMarker.detect(fileName);
+            currentImage = ImageIO.read(new File(resultFileName));
+            ImageIO.write(currentImage, "png", clientSocket.getOutputStream());
```
There is no need to send back the whole image, you should only send detected markers locations (and sizes).

alyhasansakr commented:
As you can see in the comments above, we need a little bit of restructuring:
1. There should be a separate application that generates a marker image (or video) and writes it to a file.
1. The client should read this previously generated file and send it to the server.
1. The server should detect the locations (and sizes) of the markers and send them (not the image) to the client.
1. Finally the client uses this information to modify the image.
1. Add an option for the client to show the annotated image in real time.
1. Image processing part should be done in one class that is independent of the application (image write and read should not be in the same function as detection and generation), this class should be used in all of the other 3 components (marker generator, client and server).
Let me know if you have any other questions.

Also we don't need to implement the master connection just yet, however, other parts you mentioned would be welcome.

theodorelu27 commented:
@alyhasansakr thank you for the feedbacks, I've made some progress with it but still have a long way to go
so what I've done so far was:
- Separate MarkerGenerator into a separate application, 
- Separate ImageProcessing
- Update Client to reflect the new changes

what hasn't been done:
- Send informations (detected corners and ids) instead of full image to the client) 
I expect this to be done tomorrow so test with the previous version can be done first 
My approach would be to convert both value as a JSON and stream it back to the client which decodes the received JSON and draw the detected marker from the rebuilt information
- Video processing
- Real-time annotation with options 
so real time in this context would be to update the currently displayed image/video once the detected marker is available?
- Direct execution after build from Dockerfile (currently the CMD is disabled)
- Thread execution
- Sending files from generator to client 
so in this case the ImageGenerator wouild be a client and OpenCVClient would be a server
what I had in mind is the OpenCVClient would open a thread to only receive files and queue it
once it becomes available, the queue would be popped for processing
is this interpretation correct?
- External configuration
- Unit testing
- Configuration for cmake to reduce build time

a few notes:
- As we have now multiple classes using the OpenCV shared library which need to be called once, I currently called this in the main before anything else as a static method from DetectMarker, I'm note sure if by doing this I'm giving DetectMarker too many responsibility so alternative is calling the method as is in the main or create a class for it
- I haven't retested the Dockerfiles

I've tried to list the remaining tasks in the priority that I think is important but I'd need a feedback on the priority or if my interpretation is too complicated, specific questions about the tasks have been included in the corresponding tasks

alyhasansakr commented:
@theodorelu27 That is great, go on.
I have only one comment about "Sending files from generator to client", we don't need to do that in real-time, it is enough if the generator is ran one time only to create the files (images or video) then the files are copied to the docker image (using a dockerfile). Finally the client should just read the image files (or video file) normally, and just repeats doing so in an infinite loop.

theodorelu27 commented:
@alyhasansakr thanks for the feedback, so here's the current to-do list:
- Send informations
- Video processing
- Real-time annotation with options
- External configuration
- Unit testing
- Configuration for cmake to reduce build time

theodorelu27 commented:
@alyhasansakr I've updated the Dockerfiles and codes to prioritize testing with AdvantEdge
for now Parameter can be modified by updating the client_exe.sh and server_exe.sh
Image can be built for each by moving the Dockerfile in each directory to the parent manually or using batch script

There was a problem with the OpenCV building, I've tried running it with several configurations but it's not generating the .jar file that was needed to run the application, so for now the precompiled .jar and library is provided in Resource
Yes, this is a to-do list for me too to test the configurations, currently it's taking too long to compile the OpenCV jar

aside from that, there are no major changes to the code from our last test yet

alyhasansakr commented on src/main/java/Application/OpenCVClient/Dockerfile:
```diff
@@ -0,0 +1,21 @@
+#execute
+COPY OpenCVClient/client_exe.sh client_exe.sh
+WORKDIR /
+RUN chmod +x client_exe.sh && sync && ./client_exe.sh && rm client_exe.sh
```
In AdvantEDGE, the last line will be commented and written in another interface, this line must have the required parameters, what you are doing here is writing the parameters inside a file (client_exe.sh), which means that we cannot modify them just by modifying this line.

What you can do is to expose these parameters as parameters for client_exe.sh, so this line would be:
`RUN chmod +x client_exe.sh && sync && ./client_exe.sh && rm client_exe.sh CLIENT_ID=client13 MASTER_URL=http://host.docker.internal:4567/ BEAT_COMMAND=client/register GET_SERVER_COMMAND=client/get_node/`
This way we can use it with AdvantEDGE.

And do the same with the server.

theodorelu27 commented:
@alyhasansakr in the recent commit I've exposed the parameter in the execution script, it can be modified by:
- setting environment variable
- uses provided flags when running the script

theodorelu27 commented:
@alyhasansakr I've tried to do some fixes from our last discussion:
- updated the Dockerfile to accomodate the packaging
- moved reponsibility of Exception Handling from the utlity classes to larger classes
- change Runner extension to Runnable implementation
- separate Main and Method classes

I'm still not used to Exception handling, so I might went overboard and if there's anything else that I missed, please do point it out

juan-castrillon commented on src/main/java/Application/OpenCVServer/OpenCVServer.java:
```diff
@@ -0,0 +1,47 @@
+        try {
+            activeServer.serverThread.join();
+            activeServer.heartBeatThread.join();
+        } catch (IllegalArgumentException e) {
+            e.printStackTrace();
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
```
Here again, if all handling is the same, just either do the more general (`Exception`) or the specific ones in one statement (with OR)

juan-castrillon commented on src/main/java/Application/OpenCVServer/OpenCVServerOperator.java:
```diff
@@ -0,0 +1,164 @@
+
+    private static OpenCVServerOperator instance = null;
+
+    public OpenCVServerOperator() {
```
for a singleton this should be private

juan-castrillon commented on src/main/java/Application/OpenCVServer/OpenCVServerOperator.java:
```diff
@@ -0,0 +1,164 @@
+        return instance;
+    }
+
+    public void setupServerRunners(String args[]) throws IllegalArgumentException {
```
Again declaration in C style
```suggestion
    public void setupServerRunners(String[] args) throws IllegalArgumentException {
```

juan-castrillon commented on src/main/java/Application/OpenCVServer/OpenCVServerOperator.java:
```diff
@@ -0,0 +1,164 @@
+                        masterUrl = argument[1];
+                        break;
+                    case "BEAT_COMMAND":
+                        beatCommand = argument[1];
+                        break;
+                    case "CONNECTED":
+                        connected = Boolean.parseBoolean(argument[1]);
+                        break;
+                    case "PORT":
+                        port = Integer.parseInt(argument[1]);
+                        break;
+                    default:
+                        throw new IllegalArgumentException("Invaild argument");
+                }
+            }
+        }
+        if(serverId.trim().isEmpty()) {
+            missingParameter++;
+            missingParameterList = missingParameterList + "SERVER_ID,";
+        }
+        if(serverIp.trim().isEmpty()) {
+            missingParameter++;
+            missingParameterList = missingParameterList + "SERVER_IP,";
+        }
+        if(masterUrl.trim().isEmpty()) {
+            missingParameter++;
+            missingParameterList = missingParameterList + "MASTER_URL,";
+        }
+        if(beatCommand.trim().isEmpty()) {
+            missingParameter++;
+            missingParameterList = missingParameterList + "BEAT_COMMAND,";
+        }
+        if(port == 0) {
+            missingParameter++;
+            missingParameterList = missingParameterList + "PORT";
+        }
+        if(missingParameter > 0) {
+            throw new IllegalArgumentException("Missing Parameter: " + missingParameterList);
+        }
+
```
Same comment as in the client operator. Consider changing this

juan-castrillon commented on src/main/java/Application/OpenCVServer/OpenCVServerOperator.java:
```diff
@@ -0,0 +1,164 @@
+        }
+
+        String beatUrl = masterUrl + beatCommand;
+        String beatBody =  "{" +
+                "    \"id\": \"" + serverId + "\"," +
+                "    \"ipAddress\": \"" + serverIp + "\"," +
+                "    \"connected\": " + connected + "," +
+                "    \"port\": " + port +
+                "}";
```
Same comment as in the client operator, I believe the body construction is an internal concern of the beat

juan-castrillon commented on src/main/java/Application/OpenCVClient/OpenCVClientOperator.java:
```diff
@@ -0,0 +1,269 @@
+        String response;
+        JsonNode node;
+
+        startConnection();
```
Calling this in a loop (`ProcessingRunner`) creates a new socket everytime. I dont think thats very efficient given that data like IP and port dont change, so you will be creating the same socket every iteration

juan-castrillon commented on src/main/java/Application/OpenCVServer/OpenCVServerOperator.java:
```diff
@@ -0,0 +1,164 @@
+        ImageIO.write(currentImage, "png", new File(fileName));
+        Mat subject = ImageProcessor.getImageMat(fileName);
+
+        detector = new DetectMarker();
```
This is called in a loop, so you are creating  a new detector every time. Why not just create it once and then use it all the other times? (In the constructor for example)

juan-castrillon commented on src/main/java/Application/OpenCVClient/OpenCVClientOperator.java:
```diff
@@ -0,0 +1,269 @@
+    public void markerDetection() {
+        Mat result;
+        subject = ImageProcessor.getImageMat(fileName);
+        detector = new DetectMarker();
```
Again, called in a loop,  so a new detector is called every time

juan-castrillon commented on src/main/java/Application/OpenCVServer/OpenCVServerOperator.java:
```diff
@@ -0,0 +1,164 @@
+package Application.OpenCVServer;
+
+import java.io.*;
+import java.util.Arrays;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Vector;
+import java.net.*;
+
+import javax.imageio.ImageIO;
+import java.awt.image.BufferedImage;
+
+import Application.Utilities.DetectMarker;
+import Application.Utilities.HeartBeatRunner;
+import Application.Utilities.ImageProcessor;
+import Application.Utilities.OpenCVUtil;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import org.opencv.core.Mat;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ObjectNode;
```
can be optimized

juan-castrillon commented on src/main/java/Application/OpenCVServer/OpenCVServerOperator.java:
```diff
@@ -0,0 +1,164 @@
+        fileName= "read.png";
+    }
+
+    public void startConnection() throws IOException {
+        serverSocket = new ServerSocket(port);
+        clientSocket = serverSocket.accept();
+        out = new PrintWriter(clientSocket.getOutputStream(), true);
+        in = new DataInputStream(clientSocket.getInputStream());
+    }
```
Maybe separate this behavior? Because this is also called in a loop, and in this loop:
- You want `.accept()` to be called every time because this would block until a connection is made
- You **DONT** want to create the server socket every time in the same port (just create it once)

juan-castrillon commented on src/main/java/Application/OpenCVServer/OpenCVServerOperator.java:
```diff
@@ -0,0 +1,164 @@
+        in.close();
+        out.close();
+        clientSocket.close();
+        serverSocket.close();
```
Again, do you wanna close the server socket every iteration? Or just at the end?

alyhasansakr replied:
Of course not every iteration, that would not be good for latency.

juan-castrillon commented on src/main/java/Application/OpenCVServer/ServerRunner.java:
```diff
@@ -0,0 +1,35 @@
+                activeOperator.startConnection();
+                activeOperator.standbyForConnection();
+                activeOperator.stopConnection();
+            } catch (IllegalArgumentException e) {
+                e.printStackTrace();
+            }  catch (InvalidObjectException e) {
+                e.printStackTrace();
+            }  catch (JsonProcessingException e) {
+                e.printStackTrace();
+            }  catch (IOException e) {
+                e.printStackTrace();
+            } catch (Exception e) {
+                e.printStackTrace();
```
Check the other comments about exceptions.

juan-castrillon commented on src/main/java/Application/OpenCVServer/ServerRunner.java:
```diff
@@ -0,0 +1,35 @@
+    }
+    @Override
+    public void run() {
+        while(true) {
```
Check out the comments on the ServerOperator to maybe modify this loop (dont do everything inside). Also, have in mind this loop can be exited so the thread woul never end (join) unless an exception happens or is killed

alyhasansakr replied:
This is a Docker application, so no need for this.
But anyways just add a TODO here

juan-castrillon commented on src/main/java/Application/OpenCVServer/Dockerfile:
```diff
@@ -0,0 +1,20 @@
+#execute
+COPY OpenCVServer/server_exe.sh server_exe.sh
+WORKDIR /
+RUN chmod +x server_exe.sh && sync && ./server_exe.sh && rm server_exe.sh
```
Parameters are not passed for `server.sh`

theodorelu27 replied:
this was intended because the parameters would be set as Environmental Variable in AdvantEdge and will be loaded from there

juan-castrillon reviewed (commented):
Hi @theodorelu27 ! Check out my comments and let me know if you agree or if you think I made a mistake somewhere. As final comments:

- Consider encapsulation and data-hiding based practices and check out what can be private (fields and also methods)
- Consider making some of the variables `final`
- Check out the warnings in IntelliJ. Import Optimization for example can be done automatically with one click.
- A good idea once you are done with exception handling and changing the code maybe to run a dependency analysis on IntelliJ. That you know what exactly you need for the docker containers to be as lightweight as possible

Let me know if you need any help!

theodorelu27 commented on settings.gradle:
```diff
@@ -1,2 +1,3 @@
 rootProject.name = 'Edge-Diagnostic-Platform'
 
+include ':libraries:opencv'
```
it's from when I tried to add the OpenCV LIbrary, I've removed it and it caused no problem, thanks for pointing out!

theodorelu27 commented on src/main/java/Application/OpenCVClient/OpenCVClientOperator.java:
```diff
@@ -0,0 +1,269 @@
+        Mat ids;
+        currentImage = ImageIO.read(new File(fileName));
+        ImageIO.write(currentImage, fileExtension, clientSocket.getOutputStream());
+        System.out.println("sendImage3");
```
so for this part, do you suggest to remove all or only part of the code?

juan-castrillon replied:
@theodorelu27 In this comment, I was talking about the print out. Do you need it for anything?

theodorelu27 replied:
ah, nope it was for a previous testing and it has been removed

theodorelu27 commented on src/main/java/Application/OpenCVClient/OpenCVClientOperator.java:
```diff
@@ -0,0 +1,269 @@
+        return resp;
+    }
+
+    public void stopConnection() throws IOException {
```
for this, I will add the call to stopConnection after the response has been recorded

theodorelu27 replied:
I just read you other comments, I'd address this when fixing the threads and connections

theodorelu27 commented on src/main/java/Application/OpenCVClient/OpenCVClientOperator.java:
```diff
@@ -0,0 +1,269 @@
+    public Semaphore serverLock;
+
+    //Server Communication Properties
+    private int utilizeServer;
```
it was initially a parameter received as the main argument, so I used int for more convenience
but since we've updated other input methods where this convenience is no longer relevant, I agree with you that boolean could work better

theodorelu27 commented on src/main/java/Application/OpenCVClient/OpenCVClientOperator.java:
```diff
@@ -0,0 +1,269 @@
+        try {
+            serverLock.acquire();
+            utilizeServer = input;
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        } catch (Exception e) {
+            e.printStackTrace();
```
I see, I always added the general Exception in the end in case there's anything that I might have missed, but since there is only one specific, I agree that it's more confusing to add a general exception here

as for the setServerUtilization to false, I don't think it's wise since the Exception happened when acquiring the lock, so if we modify the value without ensuring that we were able to secure the lock to do so, I think it kinda defeats the purpose
that being said I'm not too familiar with distributed systems so I would like to hear your opinion on this as well

juan-castrillon replied:
You are right about modifying the value, it beats the purpose of the lock.

Now that you mention, I wanted to ask you about the locks. If the only thing that uses this resources is a singleton, shouldn't the lock be on the singleton instance itself, not inside? 

The locks are added to ensure that only one thread can access the resource (in this case `utilizeRunner`) at the same time, and that way avoid two threads working with different values. But if the resource is modified by a singleton, by definition, there would be only one instance accessing the resources (independent from the number of threads). However, the singleton itself becomes the shared resource. So I think it would make more sense to protect it than the resources. Do you see my point? Do you agree?

theodorelu27 commented on src/main/java/Application/Utilities/DetectMarker.java:
```diff
@@ -0,0 +1,54 @@
+      this.ids = ids;
+   }
+
+   public static void initOpenCVSharedLibrary() {
```
more because it's also used in the Server
that being said, it might be better to move it to OpenCVUtil, what do you think?

juan-castrillon replied:
I agree, that class already only has static methods and it would make more sense to have it there

theodorelu27 commented on src/main/java/Application/Utilities/OpenCVUtil.java:
```diff
@@ -0,0 +1,138 @@
+
+        if(type == CvType.CV_32S || type == CvType.CV_32SC2 || type == CvType.CV_32SC3 || type == CvType.CV_16S) {
+            int[] data = new int[cols * rows * elemSize];
+            subject.get(0, 0, data);
```
it is from OpenCV Java, I'm not really sure how it works since the documentation doesn't say anything except for the parameter, but it seems the get is used to set the data field for Mat

multiple data types were used because each OpenCV type uses specific data types so I added everything to accomodate if we want to add new functionalities in the future

theodorelu27 commented on src/main/java/Application/testing_manual.md:
```diff
@@ -0,0 +1,22 @@
+``Connect the server to network - docker connect --alias [server_alias] [server_image]``<br>
+``Connect the client to network - docker connect [client_image]``
+
+Default ``[server_alias]`` is "server", you may change this in OpenCVClient.java for testing
```
this is an old manual, I'd need to rework or remove it since now we're no longer using  custom network for testing

theodorelu27 commented on src/main/resources/Configuration.json:
```diff
@@ -106,6 +106,13 @@
+      "out" : "match_m",
+      "commands" : {
+        "assign_client $client_id" : "matchMaker assign_client $client_id"
+      }
```
I've pulled the Master and there seemed to be no new changes, is there and other indicator that you can suggest to confirm that the Configuration in my current branch is up to date?

juan-castrillon replied:
For example, the match maker declaration now looks like:
```json
{"type": "MatchMaker-Random","name": "match_m"}
```

theodorelu27 replied:
okay, it should be up to date now

theodorelu27 commented:
@juan-castrillon thank you for taking the time to review my code, I agree with most of your suggestions and I've left comments on where I need your feedback on the suggestions

resolved conversations means I've included them in the recent commit and for the rest (the Exception and Threading) I'd address it later today

theodorelu27 commented on src/main/java/Application/OpenCVClient/MasterCommunicationRunner.java:
```diff
@@ -0,0 +1,52 @@
+            try {
+                System.out.println("setup server");
+                activeOperator.setup(communicator.getServer());
+            } catch (InterruptedException e) {
+                e.printStackTrace();
+            } catch (IllegalArgumentException e) {
+                e.printStackTrace();
+            } catch (ArrayIndexOutOfBoundsException e) {
+                e.printStackTrace();
+            } catch (IOException e) {
+                e.printStackTrace();
+            } catch (Exception e) {
+                e.printStackTrace();
+            }
```
yes for now we only print StackTrace
evaluation is still an abstract class, but if from the beginning the connection is not good, then it'd set ServerUtilization to false then contiue the loop to retry getting the server info

theodorelu27 commented on src/main/java/InfrastructureManager/EdgeNode.java:
```diff
@@ -4,17 +4,20 @@
     private String id;
     private String ipAddress;
     private boolean connected;
+    private int port;
```
yeah, I agree if it's used as ip:port in all other places then it might be more sensible to let them be and separate them in the application as it is a specific requirement within the application

theodorelu27 commented:
@juan-castrillon thanks for the review, I think I've addressed all changes that I think is required
as for the infinite loop, it is due to Docker applications nature so for now we'd leave it be, but I've added the exit variables as a reference for future members if they're adapting it to non Docker applications
regarding the shared lock, we want to enable other threads to modify other shared resources, so we'd leave it be for now

@alyhasansakr I think you can start the re-review

alyhasansakr commented on src/main/java/Application/OpenCVClient/MasterCommunicationRunner.java:
```diff
@@ -0,0 +1,52 @@
+
+    @Override
+    public void run() {
+        while(true) {
```
This is a Docker application, so no need for this.
But anyways just add a TODO here

alyhasansakr commented on src/main/java/Application/OpenCVClient/ProcessingRunner.java:
```diff
@@ -0,0 +1,15 @@
+    }
+    @Override
+    public void run() {
+        while(true) {
```
This is a Docker application, so no need for this.
But anyways just add a TODO here

alyhasansakr commented on src/main/java/Application/OpenCVClient/ProcessingRunner.java:
```diff
@@ -0,0 +1,19 @@
+public class ProcessingRunner implements Runnable {
+    private final OpenCVClientOperator activeOperator;
+
+    //stub for future classes
```
Add TODO:

alyhasansakr commented on src/main/java/Application/OpenCVClient/OpenCVClientOperator.java:
```diff
@@ -0,0 +1,219 @@
+    private DetectMarker detector;
+    private ObjectMapper mapper;
+
+    private final String fileName= "singlemarkerssource.png";
```
Why fixed values? at least add a TODO

alyhasansakr commented on src/main/java/Application/OpenCVClient/OpenCVClientOperator.java:
```diff
@@ -0,0 +1,219 @@
+
+    public void setupTcpConnection(String ipAddress) throws IllegalArgumentException, InterruptedException {
+        String[] argument;
+        // make a rest request to the master to ask for server and port for server
```
If this is a todo write TODO:

alyhasansakr commented on src/main/java/Application/OpenCVClient/OpenCVClientOperator.java:
```diff
@@ -0,0 +1,219 @@
+
+    public void setupClientRunners(String[] args) throws IllegalArgumentException {
+        String[] argument;
+        String clientId = ""; // for test: client13
```
Not a good place to write an example.

alyhasansakr commented on src/main/java/Application/OpenCVClient/OpenCVClientOperator.java:
```diff
@@ -0,0 +1,219 @@
+    }
+
+    public void detectMarkerInServer() throws RemoteExecutionException, IOException {
+        System.out.println("detectMarkerInServer");
```
Not a good idea to write to console in a multi-threaded application.

alyhasansakr commented on src/main/java/Application/OpenCVClient/OpenCVClientOperator.java:
```diff
@@ -0,0 +1,219 @@
+            node = mapper.readTree(response);
+            Mat ids = OpenCVUtil.deserializeMat(node.get("ids").asText());
+            List<Mat> corners = OpenCVUtil.deserializeMatList(node.get("corners").asText());
+            System.out.println("Marker Drawn with remote informations");
```
Not a good idea to write to console in a multi-threaded application.

alyhasansakr commented on src/main/java/Application/OpenCVClient/OpenCVClientOperator.java:
```diff
@@ -0,0 +1,219 @@
+
+    public void detectMarkerInClient() {
+        detector.detect(subject);
+        System.out.println("Marker Drawn locally");
```
Not a good idea to write to console in a multi-threaded application.

alyhasansakr commented on src/main/java/Application/OpenCVServer/ServerRunner.java:
```diff
@@ -0,0 +1,31 @@
+public class ServerRunner implements Runnable {
+    private final OpenCVServerOperator activeOperator;
+
+    //stub for future classes
```
Add TODO

alyhasansakr commented on src/main/java/Application/OpenCVServer/OpenCVServerOperator.java:
```diff
@@ -0,0 +1,141 @@
+    private static OpenCVServerOperator instance = null;
+
+    private OpenCVServerOperator() {
+        fileName= "read.png";
```
Why a fixed value?

alyhasansakr commented on src/main/java/Application/OpenCVServer/OpenCVServerOperator.java:
```diff
@@ -0,0 +1,141 @@
+        in = new DataInputStream(clientSocket.getInputStream());
+    }
+
+    // change name to something more appropriate
```
Add TODO:, or just change it to processing.

alyhasansakr commented on src/main/java/Application/OpenCVServer/OpenCVServerOperator.java:
```diff
@@ -0,0 +1,141 @@
+
+    public void setupServerRunners(String[] args) throws IllegalArgumentException {
+        String[] argument;
+        String masterUrl = ""; // for test: http://host.docker.internal:4567/
```
Not a good place to write examples.

alyhasansakr commented on src/main/java/Application/OpenCVServer/OpenCVServerOperator.java:
```diff
@@ -0,0 +1,141 @@
+                        beatCommand = argument[1];
+                        missingParameterList.remove("BEAT_COMMAND");
+                        break;
+                    case "CONNECTED":
```
Why is that a parameter?

theodorelu27 replied:
I put it as a parameter because beat command and get server command dependes on how it is implemented in the Master, so I wanted to give an access to change it without changing the code
I'm considering to add a TODO to move it to a config file

alyhasansakr replied:
yes, just add a TODO to move it to a config file

alyhasansakr commented on src/main/java/Application/OpenCVServer/OpenCVServerOperator.java:
```diff
@@ -0,0 +1,141 @@
+        ObjectNode resultObject = mapper.createObjectNode();
+        resultObject.put("ids", ids);
+        resultObject.put("corners", corners);
+        System.out.println("responding");
```
Printing in a multi-threaded application is not a good idea.

alyhasansakr commented on src/main/java/Application/OpenCVServer/OpenCVServerOperator.java:
```diff
@@ -0,0 +1,141 @@
+
+    // change name to something more appropriate
+    public void standbyForConnection() throws IllegalArgumentException, IOException {
+        //if input is faulty, timeout and continue
```
I don't understand that.

theodorelu27 replied:
it is an old TODO , for now it has been implemented by the Server throwing an IOException and continuing the loop

alyhasansakr commented on src/main/java/Application/OpenCVServer/OpenCVServerOperator.java:
```diff
@@ -0,0 +1,141 @@
+    public void standbyForConnection() throws IllegalArgumentException, IOException {
+        //if input is faulty, timeout and continue
+        currentImage = ImageIO.read(ImageIO.createImageInputStream(clientSocket.getInputStream()));
+        ImageIO.write(currentImage, "png", new File(fileName));
```
Give me a good reason to write to a file instead of just running the detector directly (after conversion of course).

theodorelu27 replied:
at that time I only focused on getting it to work with existing code so I wrote the file to fit it into the existing system (getting mat with filename using imread)
I just researched and I think it wouldn't take long to use the buffered image, I'd try to change it to use the Buffered Image directly

alyhasansakr commented on src/main/java/Application/OpenCVServer/ServerRunner.java:
```diff
@@ -0,0 +1,31 @@
+    public void run() {
+        while(true) {
+            try {
+                activeOperator.startConnection();
```
Do the same as you did with the client, only start a connection if not connected and close only on errors.

alyhasansakr commented on src/main/java/Application/Utilities/HeartBeatRunner.java:
```diff
@@ -0,0 +1,34 @@
+public class HeartBeatRunner implements Runnable {
+    private final EdpHeartbeat beater;
+
+    //stub for future classes
```
Add TODO

alyhasansakr commented on src/main/java/Application/Utilities/HeartBeatRunner.java:
```diff
@@ -0,0 +1,34 @@
+    }
+    @Override
+    public void run() {
+        while(true) {
```
Add TODO for use of exit and running.

alyhasansakr commented on src/main/java/Application/Utilities/EdpHeartbeat.java:
```diff
@@ -0,0 +1,44 @@
+import java.net.http.HttpResponse;
+import java.time.Duration;
+
+public class EdpHeartbeat {
```
What is "Edp"?

theodorelu27 replied:
EDP is Edge Diagnostic Platform, it's a placeholder name so can be changed if needed

alyhasansakr replied:
no problem

alyhasansakr commented on src/main/java/Application/Utilities/DetectMarker.java:
```diff
@@ -0,0 +1,46 @@
+   }
+
+   public void detect(Mat subject) {
+      Dictionary dictionary = Aruco.getPredefinedDictionary(Aruco.DICT_6X6_250);
```
Why fixed value?

theodorelu27 replied:
up until this point I only tested with that particular dictionary
but there are not particular reason so it can be taken out as a class variable or loaded from configuration (one exevution uses one library)
also is detection of markers in the same image but from multiple dictionaries within the scope of this application (one execution multiple library)?

alyhasansakr replied:
just add a TODO here

alyhasansakr reviewed (changes requested):
Other general comments:

Application files should be under src/applications/markerdetection/ and properly separated in folders java, resources, bash, docker. Also, remove files that are no longer needed (testfiles.7z for example). Make sure that everything is still working after you move the files.

As you can see there are a lot of changes that I just ask you to add a TODO instead of implementing it now. You can also do that even for the ones that I didn't specifically ask for a TODO for (unless it is essential).

For writing to console, add a parameter "debug_mode" which enables it.

Your extensive use of global variables is not a good practice, you should use more function parameters. For example, detectMarkerInClient and detectMarkerInServer should receive a Mat parameter rather than using an instance variable. Another example is making OpenCVClientOperator a Singleton for no good reason. No need to change anything here, but keep that in mind for future implementations.

theodorelu27 commented on src/main/java/Application/OpenCVClient/MasterCommunicationRunner.java:
```diff
@@ -0,0 +1,39 @@
+        while(true) {
+            try {
+                System.out.println("setup server");
+                activeOperator.setupTcpConnection("172.17.0.1:10200");
```
I was testing and forgot to revert it back, I've fixed it with its appropriate value (from MasterCommunicator's getServer)

theodorelu27 commented:
@alyhasansakr so for the feedback on runners, the TODO is to change the current implementation to not rely on while for looping?

alyhasansakr commented:
@theodorelu27 still use a while loop but with a public stop function, either implement it or write a TODO.

theodorelu27 commented on src/main/java/Application/OpenCVServer/OpenCVServerOperator.java:
```diff
@@ -32,47 +29,59 @@
     public void processing() throws IOException, IllegalArgumentException {
         currentImage = ImageIO.read(ImageIO.createImageInputStream(clientSocket.getInputStream()));
-        Mat subject = ImageProcessor.getBufferedImageMat(currentImage);
+        if(currentImage != null) {
```
@alyhasansakr @juan-castrillon 
I'm having a problem here, when server starts before the client
if the client is connected and then fails, the server didn't immediately recognize that the connection is gone so it continues the loop with the BufferedImage being null, so when client tries to reconnect

the same does not apply when client is started first however, if Client started first and I forced it to stop, there's no problem in reconnection

I'm still researching why but any suggestions would be welcomed

alyhasansakr replied:
@theodorelu27 Normally the server should be the one that always waits for the client, for example the function "accept" and "read" should block the server until a client is connected. The target is to make the server autonomous, regardless of the behavior of the client, the server must always work. And that includes recognizing that the client doesn't exist or doesn't respond.

Try one more time, if still can't do it, call me and we do it together.

theodorelu27 replied:
@alyhasansakr  I was able to reliably stop the connection when Client fails by sending an urgent message when the input stream returns null, thus throwing a Broken Pipe Exception when Client fails

so it's basically sending a byte as a heartbeat
